pkg/artifactory/repositories.go

```diff
@@ -330,442 +332,531 @@ var repoTypesLikeGeneric = []string{
-	},
-	"package_type": {
-		Type:     schema.TypeString,
-		Required: false,
-		Computed: true,
-		ForceNew: true,
-	},
-	"description": {
-		Type:     schema.TypeString,
-		Optional: true,
-	},
-	"notes": {
-		Type:     schema.TypeString,
-		Optional: true,
-	},
-	"includes_pattern": {
-		Type:        schema.TypeString,
-		Optional:    true,
-		Computed:    true,
-		Description: "List of artifact patterns to include when evaluating artifact requests in the form of x/y/**/z/*. When used, only artifacts matching one of the include patterns are served. By default, all artifacts are included (**/*).",
-	},
-	"excludes_pattern": {
-		Type:        schema.TypeString,
-		Optional:    true,
-		Computed:    true,
-		Description: "List of artifact patterns to exclude when evaluating artifact requests, in the form of x/y/**/z/*. By default no artifacts are excluded.",
-	},
-	"repo_layout_ref": {
-		Type:        schema.TypeString,
-		Optional:    true,
-		Computed:    true,
+func getBaseLocalRepoSchema(packageType string) map[string]*schema.Schema {
+
+	return map[string]*schema.Schema{
+		"key": {
+			Type:         schema.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: repoKeyValidator,
+		Description:  "A mandatory identifier for the repository that must be unique. It cannot begin with a number or contain spaces or special characters.",},
+		"project_key": {
+			Type:             schema.TypeString,
+			Optional:         true,
+			ValidateDiagFunc: projectKeyValidator,
+			Description:      "Project key for assigning this repository to. When assigning repository to a project, repository key must be prefixed with project key, separated by a dash.",
+		},
+		"project_environments": {
+			Type:        schema.TypeSet,
+			Elem:        &schema.Schema{Type: schema.TypeString},
+			MinItems:    1,
+			MaxItems:    2,
+			Set:         schema.HashString,
+			Optional:    true,
+			Description: `Project environment for assigning this repository to. Allow values: "DEV" or "PROD"`,
+		},
+		"package_type": {
+			Type:     schema.TypeString,
+			Required: false,
+			Computed: true,
+			ForceNew: true,
+		},
+		"description": {
+			Type:     schema.TypeString,
+			Optional: true,
+		},
+		"notes": {
+			Type:     schema.TypeString,
+			Optional: true,
+		},
+		"includes_pattern": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,Description: "List of artifact patterns to include when evaluating artifact requests in the form of x/y/**/z/*. When used, only artifacts matching one of the include patterns are served. By default, all artifacts are included (**/*).",
+		},
+		"excludes_pattern": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		Description: "List of artifact patterns to exclude when evaluating artifact requests, in the form of x/y/**/z/*. By default no artifacts are excluded.",},
+		"repo_layout_ref": {
```

Comment:
Having a whole function for this is kind of overkill - you can chain merges together

```go
skeema := mergeSchema(baseLocalRepo,repoLayout,whatever,whatever2)
```
The only difference is this one field and now there's a lot of changed lines.

Reply:
Improved as suggested